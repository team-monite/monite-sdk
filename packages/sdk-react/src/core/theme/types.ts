import { components } from '@/api';
import { ChipProps } from '@mui/material';

type TypographyStyle = {
  fontSize?: string | number;
  fontWeight?: string | number;
  lineHeight?: string | number;
  textTransform?: 'uppercase' | 'lowercase' | 'capitalize';
};

export type ThemeConfig = {
  borderRadius?: number;
  spacing?: number;

  colors?: {
    primary?: string;
    primaryForeground?: string;
    secondary?: string;
    neutral?: string;

    info?: string;
    success?: string;
    warning?: string;
    error?: string;

    background?: string;

    text?: string;
  };

  typography?: {
    fontFamily?: string;
    fontSize?: number;

    h1?: TypographyStyle;
    h2?: TypographyStyle;
    h3?: TypographyStyle;
    subtitle1?: TypographyStyle;
    subtitle2?: TypographyStyle;
    body1?: TypographyStyle;
    body2?: TypographyStyle;
  };

  components?: {
    invoiceStatusChip?: Partial<MoniteInvoiceStatusChipProps>;
    payableStatusChip?: Partial<MonitePayableStatusChipProps>;
    approvalRequestStatusChip?: Partial<MoniteApprovalRequestStatusChipProps>;
    invoiceRecurrenceStatusChip?: Partial<MoniteInvoiceRecurrenceStatusChipProps>;
    invoiceRecurrenceIterationStatusChip?: Partial<MoniteInvoiceRecurrenceIterationStatusChipProps>;
    counterpartStatusChip?: Partial<MoniteCounterpartStatusChipProps>;
    approvalStatusChip?: Partial<MoniteApprovalStatusChipProps>;
  };
};

export type MonitePalette = {
  primary: {
    dark: string;
    main: string;
    light: string;
    '30': string;
    '40': string;
    '50': string;
    '60': string;
    '80': string;
    '90': string;
  };
  secondary: {
    main: string;
    dark: string;
  };
  neutral: {
    '10': string;
    '50': string;
    '70': string;
    '80': string;
    '90': string;
    '95': string;
  };
  info: {
    main: string;
    light: string;
    dark: string;
  };
  success: {
    main: string;
    light: string;
    dark: string;
  };
  warning: {
    main: string;
    light: string;
    dark: string;
  };
  error: {
    main: string;
    light: string;
    dark: string;
  };
  background: {
    default: string;
    paper: string;
  };
  text: {
    primary: string;
    secondary: string;
    disabled: string;
  };
  divider: string;
  status: {
    draft: string;
    new: string;
    approve_in_progress: string;
    paid: string;
    waiting_to_be_paid: string;
    rejected: string;
    partially_paid: string;
    canceled: string;
    all: string;
  };
};

interface BaseChipProps {
  /** The variant of the Chip. */
  variant?: ChipProps['variant'];
  /** The size of the Chip. */
  size?: ChipProps['size'];
}

interface BaseStatusChipProps extends BaseChipProps {
  /** Display status icon? Set to false for dot, true for full icon, null to hide completely */
  icon?: boolean | null;
}

export type MoniteCounterpartType = 'customer' | 'vendor';

export interface MoniteInvoiceStatusChipProps extends BaseStatusChipProps {
  /** The status of the invoice. */
  status: components['schemas']['ReceivablesStatusEnum'];
}

export interface MoniteInvoiceRecurrenceStatusChipProps
  extends BaseStatusChipProps {
  /** The status of the recurrence. */
  status: components['schemas']['RecurrenceStatus'];
}

export interface MoniteInvoiceRecurrenceIterationStatusChipProps
  extends BaseStatusChipProps {
  /** The status of the iteration. */
  status: components['schemas']['IterationStatus'];
}

export interface MonitePayableStatusChipProps extends BaseStatusChipProps {
  /** The status of the payable. */
  status: components['schemas']['PayableStateEnum'];
  /** Custom colors for each payable status */
  colors?: Partial<Record<components['schemas']['PayableStateEnum'], string>>;
}

export interface MoniteCounterpartStatusChipProps extends BaseChipProps {
  /** The status of the invoice. */
  status: MoniteCounterpartType;
}

export interface MoniteApprovalRequestStatusChipProps
  extends BaseStatusChipProps {
  /** The status of the approval request. */
  status: components['schemas']['ApprovalRequestStatus'];
}

export interface MoniteApprovalStatusChipProps extends BaseStatusChipProps {
  /** The status of the payable. */
  status: components['schemas']['ApprovalPolicyStatus'];
}
