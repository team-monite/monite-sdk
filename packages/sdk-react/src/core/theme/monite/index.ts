/* eslint-disable lingui/no-unlocalized-strings */
import { ThemeConfig } from '@/core/theme/types';

export const getTheme = (theme: ThemeConfig) => {
  return {
    borderRadius: theme.borderRadius ?? 6,
    buttonBorderRadius: 8,
    spacing: theme.spacing ?? 8,

    colors: {
      primary: theme.colors?.primary || '#3737FF',
      primaryForeground: theme.colors?.primaryForeground || '#ffffff',
      secondary: theme.colors?.secondary || '#707070',
      neutral: theme.colors?.neutral || '#707070',

      info: theme.colors?.info || theme.colors?.primary || '#3737FF',
      success: theme.colors?.success || '#1FBCA0',
      warning: theme.colors?.warning || '#C78032',
      error: theme.colors?.error || '#CC394B',

      background: theme.colors?.background || '#FFFFFF',

      text: theme.colors?.text || '#292929',
    },

    typography: {
      fontFamily:
        theme?.typography?.fontFamily ||
        '"Inter", -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif',
      fontSize: theme?.typography?.fontSize ?? 16,

      h1: {
        fontSize: theme.typography?.h1?.fontSize ?? 48,
        fontWeight: theme.typography?.h1?.fontWeight ?? 600,
        lineHeight: theme.typography?.h1?.lineHeight ?? '68px',
        textTransform: theme.typography?.h1?.textTransform ?? 'none',
      },
      h2: {
        fontSize: theme.typography?.h2?.fontSize ?? 34,
        fontWeight: theme.typography?.h2?.fontWeight ?? 640,
        lineHeight: theme.typography?.h2?.lineHeight ?? '40px',
        textTransform: theme.typography?.h2?.textTransform ?? 'none',
      },
      h3: {
        fontSize: theme.typography?.h3?.fontSize ?? 24,
        fontWeight: theme.typography?.h3?.fontWeight ?? 600,
        lineHeight: theme.typography?.h3?.lineHeight ?? '32px',
        textTransform: theme.typography?.h3?.textTransform ?? 'none',
      },
      subtitle1: {
        fontSize: theme.typography?.subtitle1?.fontSize ?? 20,
        fontWeight: theme.typography?.subtitle1?.fontWeight ?? 600,
        lineHeight: theme.typography?.subtitle1?.lineHeight ?? '24px',
        textTransform: theme.typography?.subtitle1?.textTransform ?? 'none',
      },
      subtitle2: {
        fontSize: theme.typography?.subtitle2?.fontSize ?? 18,
        fontWeight: theme.typography?.subtitle2?.fontWeight ?? 600,
        lineHeight: theme.typography?.subtitle2?.lineHeight ?? '24px',
        textTransform: theme.typography?.subtitle2?.textTransform ?? 'none',
      },
      body1: {
        fontSize: theme.typography?.body1?.fontSize ?? 16,
        fontWeight: theme.typography?.body1?.fontWeight ?? 500,
        lineHeight: theme.typography?.body1?.lineHeight ?? '24px',
      },
      body2: {
        fontSize: theme.typography?.body2?.fontSize ?? 14,
        fontWeight: theme.typography?.body2?.fontWeight ?? 400,
        lineHeight: theme.typography?.body2?.lineHeight ?? '20px',
      },
    },

    components: {
      invoiceStatusChip: {
        icon: theme.components?.invoiceStatusChip?.icon || false,
        size: theme.components?.invoiceStatusChip?.size || 'small',
        variant: theme.components?.invoiceStatusChip?.variant || 'filled',
      },
      payableStatusChip: {
        icon:
          theme.components?.payableStatusChip?.icon !== undefined
            ? theme.components?.payableStatusChip?.icon
            : false,
        size: theme.components?.payableStatusChip?.size || 'small',
        variant: theme.components?.payableStatusChip?.variant || 'filled',
        colors: theme.components?.payableStatusChip?.colors,
      },
      approvalRequestStatusChip: {
        icon: theme.components?.approvalRequestStatusChip?.icon || false,
        size: theme.components?.approvalRequestStatusChip?.size || 'small',
        variant:
          theme.components?.approvalRequestStatusChip?.variant || 'filled',
      },
      invoiceRecurrenceStatusChip: {
        icon: theme.components?.invoiceRecurrenceStatusChip?.icon || false,
        size: theme.components?.invoiceRecurrenceStatusChip?.size || 'small',
        variant:
          theme.components?.invoiceRecurrenceStatusChip?.variant || 'filled',
      },
      invoiceRecurrenceIterationStatusChip: {
        icon:
          theme.components?.invoiceRecurrenceIterationStatusChip?.icon || false,
        size:
          theme.components?.invoiceRecurrenceIterationStatusChip?.size ||
          'small',
        variant:
          theme.components?.invoiceRecurrenceIterationStatusChip?.variant ||
          'filled',
      },
      counterpartStatusChip: {
        size: theme.components?.counterpartStatusChip?.size || 'small',
        variant: theme.components?.counterpartStatusChip?.variant || 'filled',
      },
      approvalStatusChip: {
        icon: theme.components?.approvalStatusChip?.icon || false,
        size: theme.components?.approvalStatusChip?.size || 'small',
        variant: theme.components?.approvalStatusChip?.variant || 'filled',
      },
    },
  };
};
