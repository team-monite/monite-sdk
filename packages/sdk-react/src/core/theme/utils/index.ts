export * from './buttonStyleHelpers';
export * from './variantMap';
export * from './withPayablesStyles';
