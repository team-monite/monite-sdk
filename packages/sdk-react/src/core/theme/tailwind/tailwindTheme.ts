// @ts-expect-error - This is a global css file
import tailwindApp from '../app.css';
import { MoniteTheme } from '@/core/context/MoniteContext';
import { ButtonStyleConfig } from '@/core/theme/types';
import { css } from '@emotion/react';

/**
 * Generates CSS custom properties for base button styles
 * @param prefix - The CSS variable prefix (e.g., 'payables-primary')
 * @param config - The button style configuration
 * @param vars - Array to push CSS variables into
 */
/* eslint-disable lingui/no-unlocalized-strings */
const generateBaseStyleVars = (
  prefix: string,
  config: ButtonStyleConfig,
  vars: string[]
): void => {
  if (config.background) {
    vars.push(`--mtw-btn-${prefix}-bg: ${config.background};`);
  }
  if (config.color) {
    vars.push(`--mtw-btn-${prefix}-color: ${config.color};`);
  }
  if (config.border) {
    vars.push(`--mtw-btn-${prefix}-border: ${config.border};`);
  }
  if (config.borderRadius !== undefined) {
    const radius =
      typeof config.borderRadius === 'number'
        ? `${config.borderRadius}px`
        : config.borderRadius;
    vars.push(`--mtw-btn-${prefix}-radius: ${radius};`);
  }
  if (config.fontWeight !== undefined) {
    vars.push(`--mtw-btn-${prefix}-weight: ${config.fontWeight};`);
  }
  if (config.boxShadow !== undefined) {
    vars.push(`--mtw-btn-${prefix}-shadow: ${config.boxShadow};`);
  }
};

/**
 * Generates CSS custom properties for button transition styles
 * @param prefix - The CSS variable prefix (e.g., 'payables-primary')
 * @param config - The button style configuration
 * @param vars - Array to push CSS variables into
 */
const generateTransitionStyleVars = (
  prefix: string,
  config: ButtonStyleConfig,
  vars: string[]
): void => {
  if (config.transitionDuration !== undefined) {
    vars.push(`--mtw-btn-${prefix}-transition-duration: ${config.transitionDuration};`);
  }
  if (config.transitionTimingFunction !== undefined) {
    vars.push(`--mtw-btn-${prefix}-transition-timing: ${config.transitionTimingFunction};`);
  }
  if (config.transitionProperty !== undefined) {
    vars.push(`--mtw-btn-${prefix}-transition-property: ${config.transitionProperty};`);
  }
};

/**
 * Generates CSS custom properties for a specific button state
 * @param prefix - The CSS variable prefix (e.g., 'payables-primary')
 * @param state - The state name ('hover', 'active', 'focus', 'disabled')
 * @param stateConfig - The state-specific configuration
 * @param vars - Array to push CSS variables into
 */
const generateStateStyleVars = (
  prefix: string,
  state: 'hover' | 'active' | 'focus' | 'disabled',
  stateConfig: Partial<ButtonStyleConfig> | undefined,
  vars: string[]
): void => {
  if (!stateConfig) return;

  if (stateConfig.background) {
    vars.push(`--mtw-btn-${prefix}-${state}-bg: ${stateConfig.background};`);
  }
  if (stateConfig.color) {
    vars.push(`--mtw-btn-${prefix}-${state}-color: ${stateConfig.color};`);
  }
  if (stateConfig.boxShadow !== undefined) {
    vars.push(`--mtw-btn-${prefix}-${state}-shadow: ${stateConfig.boxShadow};`);
  }
};

/**
 * Helper to generate CSS custom properties for button styles
 * @param prefix - The CSS variable prefix (e.g., 'payables-primary')
 * @param config - The button style configuration
 * @returns CSS string with custom properties
 */
const generateButtonVars = (prefix: string, config?: ButtonStyleConfig): string => {
  if (!config) return '';

  const vars: string[] = [];

  generateBaseStyleVars(prefix, config, vars);
  generateTransitionStyleVars(prefix, config, vars);
  generateStateStyleVars(prefix, 'hover', config.hover, vars);
  generateStateStyleVars(prefix, 'active', config.active, vars);
  generateStateStyleVars(prefix, 'focus', config.focus, vars);
  generateStateStyleVars(prefix, 'disabled', config.disabled, vars);

  return vars.map(v => `    ${v}`).join('\n');
};
/* eslint-enable lingui/no-unlocalized-strings */

/**
 * Sets the Tailwind Theme variables to the :root element
 * Based on the MoniteTheme object
 * @param theme - The MoniteTheme object
 * @uses tailwindApp - The Tailwind App CSS file
 * @returns The Tailwind Theme variables
 */
export const getTailwindTheme = (theme: MoniteTheme) => {
  const buttonStyles = theme.components?.styles?.payables?.button;

  return css`
    :root,
    :host {
      --mtw-font-family: ${theme.typography.fontFamily};

    --mtw-color-primary-10: ${theme.palette.primary[10]};
    --mtw-color-primary-20: ${theme.palette.primary[20]};
    --mtw-color-primary-30: ${theme.palette.primary[30]};
    --mtw-color-primary-40: ${theme.palette.primary[40]};
    --mtw-color-primary-50: ${theme.palette.primary[50]};
    --mtw-color-primary-55: ${theme.palette.primary[55]};
    --mtw-color-primary-60: ${theme.palette.primary[60]};
    --mtw-color-primary-65: ${theme.palette.primary[65]};
    --mtw-color-primary-80: ${theme.palette.primary[80]};
    --mtw-color-primary-85: ${theme.palette.primary[85]};
    --mtw-color-primary-90: ${theme.palette.primary[90]};
    --mtw-color-primary-95: ${theme.palette.primary[95]};
    --mtw-color-primary-foreground: ${theme.palette.primary.foreground?.main || theme.palette.primary.contrastText};

    --mtw-color-neutral-10: ${theme.palette.neutral[10]};
    --mtw-color-neutral-30: ${theme.palette.neutral[30]};
    --mtw-color-neutral-50: ${theme.palette.neutral[50]};
    --mtw-color-neutral-70: ${theme.palette.neutral[70]};
    --mtw-color-neutral-80: ${theme.palette.neutral[80]};
    --mtw-color-neutral-90: ${theme.palette.neutral[90]};
    --mtw-color-neutral-95: ${theme.palette.neutral[95]};

    --mtw-color-danger-10: ${theme.palette.error[10]};
    --mtw-color-danger-25: ${theme.palette.error[25]};
    --mtw-color-danger-50: ${theme.palette.error[50]};
    --mtw-color-danger-75: ${theme.palette.error[75]};
    --mtw-color-danger-100: ${theme.palette.error[100]};

    --mtw-color-success-10: ${theme.palette.success[10]};
    --mtw-color-success-30: ${theme.palette.success[30]};
    --mtw-color-success-50: ${theme.palette.success[50]};
    --mtw-color-success-60: ${theme.palette.success[60]};
    --mtw-color-success-80: ${theme.palette.success[80]};
    --mtw-color-success-90: ${theme.palette.success[90]};
    --mtw-color-success-95: ${theme.palette.success[95]};

    --mtw-color-warning-10: ${theme.palette.warning[10]};
    --mtw-color-warning-30: ${theme.palette.warning[30]};
    --mtw-color-warning-50: ${theme.palette.warning[50]};
    --mtw-color-warning-60: ${theme.palette.warning[60]};
    --mtw-color-warning-80: ${theme.palette.warning[80]};
    --mtw-color-warning-90: ${theme.palette.warning[90]};
    --mtw-color-warning-95: ${theme.palette.warning[95]};

${generateButtonVars('payables-primary', buttonStyles?.primary)}
${generateButtonVars('payables-secondary', buttonStyles?.secondary)}
${generateButtonVars('payables-tertiary', buttonStyles?.tertiary)}
${generateButtonVars('payables-destructive', buttonStyles?.destructive)}
  }

  ${tailwindApp}
`
};
