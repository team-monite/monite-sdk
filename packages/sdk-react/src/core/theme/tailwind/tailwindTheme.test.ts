import { generateThemeCssVars } from './tailwindTheme';
import { MoniteTheme } from '@/core/context/MoniteContext';
import { ButtonStyleConfig, ThemeConfig } from '@/core/theme/types';
import { createThemeWithDefaults } from '@/core/utils/createThemeWithDefaults';

/**
 * Tests for CSS variable generation.
 *
 * ## Testing Strategy
 *
 * We test `generateThemeCssVars()` directly, which contains all the business logic
 * for extracting theme values as CSS custom properties. `getTailwindTheme()` is a
 * thin wrapper that passes this output to emotion's `css` template literal.
 *
 * ### Why We Don't Test getTailwindTheme() Directly
 *
 * Testing `getTailwindTheme()` in Jest presents challenges:
 *
 * 1. **Emotion's Internal Serialization**: The `css` template tag uses emotion's
 *    internal serialization (`@emotion/serialize`) which calls `.toString()` on
 *    interpolated values. Mocking `css` alone doesn't prevent this serialization.
 *
 * 2. **CSS Import Complications**: `tailwindApp` is imported from a CSS file, which
 *    Jest's `identity-obj-proxy` turns into a Proxy object that emotion can't serialize.
 *
 * 3. **Complex Mocking Required**: To properly mock emotion, we'd need to mock:
 *    - `@emotion/react`'s `css` function
 *    - `@emotion/serialize`'s serialization logic
 *    - The CSS import from `../app.css`
 *
 *    This level of mocking tests our mocks more than our code.
 *
 * ### What We Test Instead
 *
 * By testing `generateThemeCssVars()` directly, we validate:
 * - ✅ All theme colors are extracted correctly
 * - ✅ Button customizations generate proper CSS variables
 * - ✅ CSS syntax is correct (semicolons, colons, values)
 * - ✅ State-specific variables (hover, active, focus, disabled) are generated
 * - ✅ All button variants (primary, secondary, tertiary, destructive) work
 *
 */

const createTestTheme = (
  buttonOverrides?: {
    primary?: ButtonStyleConfig;
    secondary?: ButtonStyleConfig;
    tertiary?: ButtonStyleConfig;
    destructive?: ButtonStyleConfig;
  }
): MoniteTheme => {
  const themeConfig: ThemeConfig = {
    components: buttonOverrides
      ? {
          styles: {
            payables: {
              button: buttonOverrides,
            },
          },
        }
      : undefined,
  };

  return createThemeWithDefaults(themeConfig);
};

describe('tailwindTheme CSS Variable Generators', () => {
  describe('generateThemeCssVars', () => {
    test('generates CSS with base color variables', () => {
      const theme = createTestTheme();
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-color-primary-10');
      expect(result).toContain('--mtw-color-neutral-50');
      expect(result).toContain('--mtw-color-danger-50');
      expect(result).toContain('--mtw-font-family');
    });

    test('generates button CSS variables when config is provided', () => {
      const buttonConfig: ButtonStyleConfig = {
        background: '#667eea',
        color: '#ffffff',
        border: '2px solid #667eea',
        borderRadius: 12,
        fontWeight: 600,
        boxShadow: '0 4px 6px rgba(0, 0, 0, 0.1)',
      };

      const theme = createTestTheme({ primary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-primary-bg: #667eea');
      expect(result).toContain('--mtw-btn-payables-primary-color: #ffffff');
      expect(result).toContain('--mtw-btn-payables-primary-border: 2px solid #667eea');
      expect(result).toContain('--mtw-btn-payables-primary-radius: 12px');
      expect(result).toContain('--mtw-btn-payables-primary-weight: 600');
      expect(result).toContain('--mtw-btn-payables-primary-shadow: 0 4px 6px rgba(0, 0, 0, 0.1)');
    });

    test('generates hover state CSS variables', () => {
      const buttonConfig: ButtonStyleConfig = {
        background: '#667eea',
        hover: {
          background: '#5568d3',
          color: '#f0f0f0',
          boxShadow: '0 6px 12px rgba(0, 0, 0, 0.2)',
        },
      };

      const theme = createTestTheme({ primary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-primary-hover-bg: #5568d3');
      expect(result).toContain('--mtw-btn-payables-primary-hover-color: #f0f0f0');
      expect(result).toContain('--mtw-btn-payables-primary-hover-shadow: 0 6px 12px rgba(0, 0, 0, 0.2)');
    });

    test('generates all state CSS variables (hover, active, focus, disabled)', () => {
      const buttonConfig: ButtonStyleConfig = {
        background: '#667eea',
        hover: { background: '#5568d3' },
        active: { background: '#4557bd' },
        focus: { background: '#7788ee' },
        disabled: {
          background: '#e5e7eb',
          color: '#9ca3af',
        },
      };

      const theme = createTestTheme({ secondary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-secondary-hover-bg: #5568d3');
      expect(result).toContain('--mtw-btn-payables-secondary-active-bg: #4557bd');
      expect(result).toContain('--mtw-btn-payables-secondary-focus-bg: #7788ee');
      expect(result).toContain('--mtw-btn-payables-secondary-disabled-bg: #e5e7eb');
      expect(result).toContain('--mtw-btn-payables-secondary-disabled-color: #9ca3af');
    });

    test('generates variables for all button variants', () => {
      const theme = createTestTheme({
        primary: { background: '#667eea' },
        secondary: { background: '#ffffff' },
        tertiary: { background: 'transparent' },
        destructive: { background: '#ef4444' },
      });

      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-primary-bg: #667eea');
      expect(result).toContain('--mtw-btn-payables-secondary-bg: #ffffff');
      expect(result).toContain('--mtw-btn-payables-tertiary-bg: transparent');
      expect(result).toContain('--mtw-btn-payables-destructive-bg: #ef4444');
    });

    test('generates transition CSS variables', () => {
      const buttonConfig: ButtonStyleConfig = {
        background: '#667eea',
        transitionDuration: '200ms',
        transitionTimingFunction: 'ease-in-out',
        transitionProperty: 'all',
      };

      const theme = createTestTheme({ primary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-primary-transition-duration: 200ms');
      expect(result).toContain('--mtw-btn-payables-primary-transition-timing: ease-in-out');
      expect(result).toContain('--mtw-btn-payables-primary-transition-property: all');
    });

    test('handles gradient backgrounds correctly', () => {
      const buttonConfig: ButtonStyleConfig = {
        background: 'linear-gradient(135deg, #667eea 0%, #764ba2 100%)',
        hover: {
          background: 'linear-gradient(135deg, #5568d3 0%, #65408b 100%)',
        },
      };

      const theme = createTestTheme({ primary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('linear-gradient(135deg, #667eea 0%, #764ba2 100%)');
      expect(result).toContain('linear-gradient(135deg, #5568d3 0%, #65408b 100%)');
    });

    test('handles borderRadius as string (CSS value)', () => {
      const buttonConfig: ButtonStyleConfig = {
        borderRadius: '0.5rem',
      };

      const theme = createTestTheme({ tertiary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-tertiary-radius: 0.5rem');
    });

    test('handles borderRadius as number (converted to px)', () => {
      const buttonConfig: ButtonStyleConfig = {
        borderRadius: 8,
      };

      const theme = createTestTheme({ destructive: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-destructive-radius: 8px');
    });

    test('handles empty button config gracefully', () => {
      const theme = createTestTheme({});
      const result = generateThemeCssVars(theme);

      expect(result).toBeDefined();
      expect(result).toBeDefined();
      expect(result).toContain('--mtw-color-primary-10');
    });

    test('handles undefined button config gracefully', () => {
      const theme = createTestTheme();

      const result = generateThemeCssVars(theme);

      expect(result).toBeDefined();
      expect(result).toBeDefined();
    });

    test('handles complex box shadow with multiple values', () => {
      const buttonConfig: ButtonStyleConfig = {
        boxShadow: '0 10px 25px rgba(0, 0, 0, 0.15), 0 4px 10px rgba(0, 0, 0, 0.08)',
      };

      const theme = createTestTheme({ primary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('0 10px 25px rgba(0, 0, 0, 0.15), 0 4px 10px rgba(0, 0, 0, 0.08)');
    });

    test('handles borderRadius of 0', () => {
      const buttonConfig: ButtonStyleConfig = {
        borderRadius: 0,
      };

      const theme = createTestTheme({ primary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-primary-radius: 0px');
    });

    test('handles fontWeight as string', () => {
      const buttonConfig: ButtonStyleConfig = {
        fontWeight: 'bold',
      };

      const theme = createTestTheme({ secondary: buttonConfig });
      const result = generateThemeCssVars(theme);

      expect(result).toContain('--mtw-btn-payables-secondary-weight: bold');
    });

    test('does not generate double semicolons', () => {
      const theme = createTestTheme({
        primary: { background: '#667eea' },
        secondary: { background: '#ffffff' },
      });

      const result = generateThemeCssVars(theme);

      expect(result).not.toContain(';;');
    });

    test('generates properly formatted CSS variables', () => {
      const theme = createTestTheme({
        primary: { background: '#667eea' },
      });
      const result = generateThemeCssVars(theme);

      const lines = result.split('\n');
      expect(lines.length).toBeGreaterThan(1);

      lines.forEach((line) => {
        const trimmed = line.trim();
        if (trimmed !== '' && trimmed !== ';') {
          expect(line).toMatch(/^--[\w-]+: .+;?$/);
        }
      });
    });

    test('adds blank line separator between color vars and button vars', () => {
      const theme = createTestTheme({
        primary: { background: '#667eea' },
      });

      const result = generateThemeCssVars(theme);
      const vars = result.split(';\n');
      const hasBlankLine = vars.some(v => v.trim() === '');

      expect(hasBlankLine).toBe(true);
    });

    test('snapshot: complete CSS output with all button variants', () => {
      const theme = createTestTheme({
        primary: {
          background: 'linear-gradient(135deg, #667eea 0%, #764ba2 100%)',
          color: '#ffffff',
          borderRadius: 12,
          fontWeight: 600,
          boxShadow: '0 4px 6px rgba(102, 126, 234, 0.3)',
          hover: {
            background: 'linear-gradient(135deg, #5568d3 0%, #65408b 100%)',
            boxShadow: '0 6px 20px rgba(102, 126, 234, 0.6)',
          },
        },
        secondary: {
          background: '#ffffff',
          color: '#667eea',
          border: '2px solid #667eea',
          borderRadius: 8,
          hover: {
            background: '#f5f7ff',
          },
        },
      });

      const result = generateThemeCssVars(theme);

      expect(result).toMatchSnapshot();
    });
  });
});
