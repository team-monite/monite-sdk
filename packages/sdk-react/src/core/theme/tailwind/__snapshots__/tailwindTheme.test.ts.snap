// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tailwindTheme CSS Variable Generators generateThemeCssVars snapshot: complete CSS output with all button variants 1`] = `
"--mtw-font-family: "Inter", -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;
--mtw-color-primary-10: #00009a;
--mtw-color-primary-20: #0000a9;
--mtw-color-primary-30: #0003cd;
--mtw-color-primary-40: #0020e4;
--mtw-color-primary-50: #3737FF;
--mtw-color-primary-55: #5245fc;
--mtw-color-primary-60: #8988fc;
--mtw-color-primary-65: #a8a7f9;
--mtw-color-primary-80: #d9d9f2;
--mtw-color-primary-85: #e7e7ef;
--mtw-color-primary-90: #ededf5;
--mtw-color-primary-95: #f3f3f3;
--mtw-color-primary-foreground: #ffffff;
--mtw-color-neutral-10: #080808;
--mtw-color-neutral-30: #313131;
--mtw-color-neutral-50: #707070;
--mtw-color-neutral-70: #909090;
--mtw-color-neutral-80: #dedede;
--mtw-color-neutral-90: #f2f2f2;
--mtw-color-neutral-95: #fafafa;
--mtw-color-danger-10: #D42623;
--mtw-color-danger-25: #DF5C5A;
--mtw-color-danger-50: #EA9391;
--mtw-color-danger-75: #F4C9C8;
--mtw-color-danger-100: #FCF1F1;
--mtw-color-success-10: #12705f;
--mtw-color-success-30: #0dac8e;
--mtw-color-success-50: #1fbca0;
--mtw-color-success-60: #4dd6bd;
--mtw-color-success-80: #ace5da;
--mtw-color-success-90: #ccf0e9;
--mtw-color-success-95: #eefbf9;
--mtw-color-warning-10: #996600;
--mtw-color-warning-30: #cc8000;
--mtw-color-warning-50: #ff9900;
--mtw-color-warning-60: #ffb333;
--mtw-color-warning-80: #ffd999;
--mtw-color-warning-90: #ffe6cc;
--mtw-color-warning-95: #fff3ee;
;
--mtw-btn-payables-primary-bg: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
--mtw-btn-payables-primary-color: #ffffff;
--mtw-btn-payables-primary-radius: 12px;
--mtw-btn-payables-primary-weight: 600;
--mtw-btn-payables-primary-shadow: 0 4px 6px rgba(102, 126, 234, 0.3);
--mtw-btn-payables-primary-hover-bg: linear-gradient(135deg, #5568d3 0%, #65408b 100%);
--mtw-btn-payables-primary-hover-shadow: 0 6px 20px rgba(102, 126, 234, 0.6);
--mtw-btn-payables-secondary-bg: #ffffff;
--mtw-btn-payables-secondary-color: #667eea;
--mtw-btn-payables-secondary-border: 2px solid #667eea;
--mtw-btn-payables-secondary-radius: 8px;
--mtw-btn-payables-secondary-hover-bg: #f5f7ff;"
`;
